### ISSUE TITLE
Remove reference + dynamic requirements from combiner generators

### ISSUE THREAD

**Issue by Lucretiel** [Razaekel/noise-rs#201]
Many of the utility generators, like `Add` and `Clamp`, could be rewritten to use more generic code. This would have the advantage of removing the need to take generators by reference, as well as the need to use dynamic dispatch; both of these would significantly improve performance and code style.

For instance, consider Add. You could rewrite it like so:

```rust
// Also, you should add some derived impls to all your structs. At the VERY least,
// you should derive Debug unconditionally.
#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
struct Add<A, B>(pub A, pub B);

impl<T: Copy, A: NoiseFn<T>, B: NoiseFn<T>> NoiseFn<T> for Add<A. B>
{
    fn get(&self, point: T) -> f64 {
        self.0.get(point) + self.1.get(point)
    }
}
```

The compiler can then substantially inline this code, because it doesn't have to go through the dynamic dispatch of using `&'a NoiseFn<T>`.

This would technically be a breaking change, because all the type signatures are different, but idiomatic use of the library should actually continue to work just fine.

**Comment by Lucretiel** [comment]
Note: I am working on this in my [fork](https://github.com/Lucretiel/noise-rs) of this repo. Virtually every generator in the library could benefit from this change.

Currently, since this constitutes a substantial breaking change (due to radically changed type signatures of all the structs), I'm also taking the opportunity to streamline a lot of the classes. Please let me know in this issue if you'd rather retain the older interfaces to these structs.

For instance, it's not really necessary to have setter methods for public fields, and the simplest combinators (like `Add`) don't need to have named fields.

**Comment by nicoxxl** [comment]
I have a similar requirement, with references, it is not possible (at least I don't know how) to return a ScaleBias of a Perlin.
I get a really nice :+1:
```
error[E0515]: cannot return value referencing temporary value


return Box::new(ScaleBias::new(&Perlin::new()).set_scale(0.5).set_bias(0.5));
                               ------------- temporary value created here
```

**Comment by Razaekel** [comment]
Closed by #256
